File: main/add-and-search-word-data-structure-design/add-and-search-word-data-structure-design-scratch.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <memory>
#include <queue>
#include <string>
#include <unordered_map>
#include <utility>

using namespace std;

class WordDictionary {
public:
    /** Initialize your data structure here. */
    WordDictionary() = default;
    WordDictionary(const WordDictionary&) = delete;
    WordDictionary(WordDictionary&&) = default;
    WordDictionary& operator=(const WordDictionary&) = delete;
    WordDictionary& operator=(WordDictionary&&) = default;
    ~WordDictionary();
    
    /** Adds a word into the data structure. */
    void addWord(string word);
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) const;
    
private:
    using Children = unordered_map<char, WordDictionary>;
    
    unique_ptr<Children> children_ {make_unique<Children>()};
    bool terminal_ {false};
};

WordDictionary::~WordDictionary() {
    if (!children_) return;
    
    queue<unique_ptr<Children>> q;
    for (q.emplace(move(children_)); !q.empty(); q.pop()) {
        for (auto& kv : *q.front()) {
            assert(kv.second.children_);
            q.emplace(move(kv.second.children_));
        }
    }
}

void WordDictionary::addWord(const string word) {
    auto p = this;
    for (const auto ch : word) p = &(*p->children_)[ch];
    p->terminal_ = true;
}

bool WordDictionary::search(const string word) const {
    queue<const WordDictionary*> q;
    q.push(this);
    
    for (const auto ch : word) {
        if (ch == '.') {
            for (auto width = q.size(); width != 0; --width) {
                const auto p = q.front();
                q.pop();
                
                // . means any character, so enqueue all the next level's nodes
                for (const auto& kv : *p->children_) q.push(&kv.second);
            }
        } else {
            for (auto width = q.size(); width != 0; --width) {
                const auto p = q.front();
                q.pop();
                
                // enqueue only the nodes that match ch literally, if any
                const auto it = p->children_->find(ch);
                if (it != p->children_->end()) q.push(&it->second);
            }
        }
        
        if (q.empty()) return false;
    }
    
    // the word matches iff any match of the last letter is terminal
    do {
        if (q.front()->terminal_) return true;
        q.pop();
    }
    while (!q.empty());
    
    return false;
}

namespace {
    void run_tests() {
        WordDictionary trie;

        const auto test = [&trie](auto word, bool show = true) {
            cout << "search(\"" << (show ? word : "<not shown>") << "\") -> \""
                 << trie.search(word) << "\"\n";
        };

        test("");
        test(".");
        test("..");
        test("...");
        cout << '\n';

        trie.addWord("abababababababababababababababababababababababababab");
        trie.addWord("ababababababababababababababababababababababcdcdcdcd");
        test(        "ababababababababababababababababababababa..........d");
        test(        "ababababababababababababababababababababa.........ad");
        cout << '\n';

        string s (500, 'y');
        trie.addWord(s);
        fill(begin(s), end(s), '.');
        test(s);

        s += '.';
        cout << s.size() << '\n';
        test(s);

        s.erase(s.size() - 2, 2);
        cout << s.size() << '\n';
        test(move(s));

        cout << s.size() << '\n';
        test(s);
        cout << '\n';

        string t (50'000, 'z');
        trie.addWord(t);
        test(t, false);
        fill(begin(t), end(t), '.');
        test(t, false);

        t += '.';
        cout << t.size() << '\n';
        test(t, false);

        t.erase(t.size() - 2, 2);
        cout << t.size() << '\n';
        test(move(t), false);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cout << boolalpha;

    run_tests();
}

File: main/add-and-search-word-data-structure-design/add-and-search-word-data-structure-design-scratch.java
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;
import java.util.function.Consumer;

class WordDictionary {
    /** Adds a word into the data structure. */
    public void addWord(final String word) {
        WordDictionary cur = this;

        for (int i = 0; i != word.length(); ++i) {
            cur = cur._children.computeIfAbsent(word.charAt(i),
                    key -> new WordDictionary());
        }

        cur._terminal = true;
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(final String word) {
        final Queue<WordDictionary> q = new ArrayDeque<>();
        q.add(this);

        for (int i = 0; i != word.length(); ++i) {
            final char ch = word.charAt(i);

            if (ch == '.') {
                for (int width = q.size(); width != 0; --width) {
                    // . means any char, so enqueue all the next level's nodes
                    q.addAll(q.remove()._children.values());
                }
            } else {
                for (int width = q.size(); width != 0; --width) {
                    // enqueue only the nodes that match ch literally, if any
                    final WordDictionary child = q.remove()._children.get(ch);
                    if (child != null) q.add(child);
                }
            }

            if (q.isEmpty()) return false; // a simple optimization
        }

        return q.stream().anyMatch(node -> node._terminal);
    }

    private final Map<Character, WordDictionary> _children = new HashMap<>();

    private boolean _terminal = false;
}

final class UnitTest {
    public static void main(final String[] args) {
        final WordDictionary trie = new WordDictionary();

        final Consumer<String> test = word -> System.out.format(
                "search(\"%s\") -> \"%s\"%n", word, trie.search(word));

        test.accept("");
        test.accept(".");
        test.accept("..");
        test.accept("...");
        System.out.println();

        trie.addWord("abababababababababababababababababababababababababab");
        trie.addWord("ababababababababababababababababababababababcdcdcdcd");
        test.accept( "ababababababababababababababababababababa..........d");
        test.accept( "ababababababababababababababababababababa.........ad");
    }
}
